### DISCUSSION TITLE
Wassertein distance

### DISCUSSION THREAD

**Discussion by ghost** [Project-MONAI/MONAI#1267] [Q&A]
Hello, I wanted to ask if anyone has used Wasserstein distance in brain different structures segmentation because I have some issues. For example, the argument that I should pass to my pipeline is a matrix distance and I would like to know how I construct this matrix, I don't know where those numbers come from. And the other question if is the loss of Wasserstein distance is finished and prove in any experiment, I only found on the internet the Wasserstein distance in brain tumor. I would like if someone knows more examples because in brain tumors the labels flow into each other. In my case, I am trying to do the Wasserstein distance between structures such as putamen, caudate, and thalamus. 

Thank you.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Could you please help with this question? I think you are an expert on these research details.

Thanks.

**Comment by LucasFidon** [comment]
Hi @natacanta5,

Thank you for your question!

The generalized Wasserstein Dice loss was indeed first proposed for brain tumor segmentation in our work:
[Fidon, Lucas, et al. "Generalised Wasserstein dice score for imbalanced multi-class segmentation using holistic convolutional networks." International MICCAI Brainlesion Workshop. Springer, Cham, 2017.](https://link.springer.com/chapter/10.1007/978-3-319-75238-9_6)

In section 5.1, we have proposed a heuristic to design the distance matrix (please see also fig. 1).
The heuristic is based on a tree for the regions to be segmented. If you can define such a tree for your application, then you can use this heuristic to define your distance matrix.

In your example with putamen, caudate, and thalamus, one possible tree would be:
![image](https://user-images.githubusercontent.com/17875992/99881581-d6d53180-2c1a-11eb-9846-ecb6542cf500.png)
Feel free to add more regions.

For more examples of applications of the generalized Wasserstein Dice loss and how they defined their distance matrix, you can look at:
- more on brain tumor segmentation from MRI
[Fidon, Lucas, Sebastien Ourselin, and Tom Vercauteren. "Generalized Wasserstein Dice Score, Distributionally Robust Deep Learning, and Ranger for brain tumor segmentation: BraTS 2020 challenge." arXiv preprint arXiv:2011.01614 (2020).](https://arxiv.org/abs/2011.01614)
- segmentation of lung lesions due to COVID-19 from CT (see the Appendix)
[Tilborghs, Sofie, et al. "Comparative study of deep learning methods for the automatic segmentation of lung, lesion and lesion type in CT scans of COVID-19 patients." arXiv preprint arXiv:2007.15546 (2020).](https://arxiv.org/abs/2007.15546)
- segmentation of subcutaneous adipose tissue, visceral adipose tissue, muscular body mass, bone, and visceral organs from low-dose CT images (see the supplementary material)
[Blanc-Durand, Paul, et al. "Prognostic value of anthropometric measures extracted from whole-body CT using deep learning in patients with non-small-cell lung cancer." European Radiology (2020): 1-10.](https://link.springer.com/article/10.1007%2Fs00330-019-06630-w)

There may be more examples in the literature that I am not aware of.

**Comment by LucasFidon** [comment]
Hi @Nic-Ma @wyli,

the current implementation of the generalized Wasserstein Dice loss in MONAI is not optimal #1272 
I am working at updating the implementation based on my (small) git repository for the generalized Wasserstein Dice loss
https://github.com/LucasFidon/GeneralizedWassersteinDiceLoss

I have noticed that the generalized Wasserstein Dice loss does not appear in the documentation of MONAI at the moment.
Should I add the documentation as part of my PR for #1272?

Thanks

**Comment by ghost** [comment]
Hello,

I have been working in the t¡matrix but i still have many doubts and I am not sure if the Wassertein distance could work in my example. In the tree you mention, for each structure there is left and right structures (putamen: left and right, caudate: left and right, thalamus: left and right). So I attach the matrix that I should define to try to implement the Wassertein distance. My principal problem is the number of the relation between each structure that you put on the tree.

<img width="809" alt="Captura de pantalla 2020-12-01 a las 18 37 34" src="https://user-images.githubusercontent.com/50491642/100776162-7e651780-3404-11eb-9d87-f1b66b58ef2d.png">

**Comment by saruarlive** [comment]
@LucasFidon, @wyli, @Nic-Ma 
Hi, hope you guys are doing good.
What is the current status of Wasserstein Dice loss class in monai library? Is it updated accordingly? The reason I am asking because I had used the loss function from Monai, it turns out that Dice+cross entropy performs better in brain tumor segmentation? Shall I update the monai from source if the loss function is optimal now?

**Comment by LucasFidon** [comment]
Hi @saruarlive, 

Thank you for your message!

I have updated the Wasserstein Dice loss in MONAI (and its documentation) (please see #1272 for more detail).
But I don't think it is part of the latest release of MONAI yet, so you may have to install MONAI from source.

Yes, please, use the latest version of the Wasserstein Dice loss for brain tumor segmentation. For the distance matrix, you can use the one we used in our [BraTS20 paper](https://arxiv.org/pdf/2011.01614.pdf).

Note that at the time of writing the [Wasserstein Dice loss paper](https://arxiv.org/pdf/1707.00478.pdf), Dice + cross entropy was not a popular choice of loss function yet.
I would advise you to compare this loss to Wasserstein Dice + cross entropy (instead of the Wasserstein Dice loss alone).
In my experiments for BraTS19 and BraTS20, I have found that Wasserstein Dice + cross entropy outperforms Dice + cross entropy and Wassertein Dice alone for different deep learning pipelines.

**Comment by LucasFidon** [comment]
Hi @natacanta5,
The entries that you have already filled in your matrix are correct.
Maybe you can start by considering only the putamen, the thalamus, and the caudate (without distinguishing right and left).
In this case, would you be able to define your distance matrix based on the tree above?
For the weights on the tree, you have to choose values between 0 and 1.

**Comment by saruarlive** [comment]
@LucasFidon,

I had read your paper before. Great work you have done. 
Would you recommend if any kind of normalization or noise-robust parameter needs to be introduced there as you had done in COVID19 CT segmenation challenge I guess? I am looking forward to hearing from you.

**Comment by LucasFidon** [comment]
Hi @saruarlive,
Thank you!
For brain tumor segmentation, I have found that the original Wasserstein Dice loss (with `weighting_mode="default"` in MONAI) worked better than the Wasserstein Dice loss with the alternative normalization that we used for COVID-19 (with `weighting_mode="GDL"` in MONAI). It was only a quick comparison, so this has to be interpreted with caution.
It seems that the BraTS manual segmentations are less noisy than the COVID-19 manual segmentations.
